### PR TITLE
some GCLowering cleanup and add support for FCAs

### PR DESCRIPTION
**Pull request by vtjnash**
In preparation for some future work, change some representations and cleanup some code.

### CODE REVIEW

File: test/llvmpasses/gcroots.ll [hunk 1]
```diff
@@ -124,11 +125,12 @@ blabel:
     %bdecayed = addrspacecast %jl_value_t addrspace(10)* %bboxed to i64 addrspace(12)*
     br label %common
 common:
-    %phi = phi i64 addrspace(12)* [ %adecayed, %alabel ], [ %bdecayed, %blabel ]
+    %phi = phi i64 addrspace(12)* [ %adecayed, %alabel ], [ %bdecayed, %blabel ], [ %phi, %common ]
     call void @one_arg_decayed(i64 addrspace(12)* %phi)
-    ret void
+    br label %common
```
Comment (vchuravy) [inline test/llvmpasses/gcroots.ll]: Why is this now a infinite for-loop?

File: src/codegen.cpp [hunk 2]
```diff
@@ -5712,7 +5719,7 @@ static std::unique_ptr<Module> emit_function(
                 varinfo.dinfo = dbuilder.createParameterVariable(
                     SP,                                 // Scope (current function will be fill in later)
                     jl_symbol_name(argname),            // Variable name
-                    ctx.has_sret + i + 1,               // Argument number (1-based)
+                    has_sret + i + 1,               // Argument number (1-based)
```
Comment (vchuravy) [inline src/codegen.cpp]: formatting of the comment

File: src/codegen.cpp [hunk 3]
```diff
@@ -5724,7 +5731,7 @@ static std::unique_ptr<Module> emit_function(
                 ctx.slots[ctx.vaSlot].dinfo = dbuilder.createParameterVariable(
                     SP,                                 // Scope (current function will be fill in later)
                     std::string(jl_symbol_name(slot_symbol(ctx, ctx.vaSlot))) + "...",  // Variable name
-                    ctx.has_sret + nreq + 1,            // Argument number (1-based)
+                    has_sret + nreq + 1,            // Argument number (1-based)
```
Comment (vchuravy) [inline src/codegen.cpp]: formatting of the comment.

File: src/ccall.cpp [hunk 4]
```diff
@@ -673,6 +673,8 @@ static jl_cgval_t emit_cglobal(jl_codectx_t &ctx, jl_value_t **args, size_t narg
         rt = (jl_value_t*)jl_voidpointer_type;
     }
     Type *lrt = julia_type_to_llvm(rt);
+    if (lrt == T_prjlvalue)
+        lrt = T_pjlvalue;
```
[the remark this answers is not in the document]
Comment (vtjnash) [reply src/ccall.cpp]: true, `assert(lrt != T_prjlvalue)` might be better

File: test/llvmpasses/gcroots.ll (hunk 1 above)
Comment (vtjnash) [reply test/llvmpasses/gcroots.ll]: easiest way to make the test more complex without affecting anything that mattered

**Comment by Keno** [comment]
Perhaps cherry-pick the `src: fix and sort makefile rules` onto master?

File: src/llvm-late-gc-lowering.cpp [hunk 5]
```diff
@@ -811,12 +804,13 @@ void LateLowerGCFrame::NoteUse(State &S, BBState &BBS, Value *V, BitVector &Uses
...
 }
 
-void LateLowerGCFrame::NoteOperandUses(State &S, BBState &BBS, User &UI, BitVector &Uses) {
+void LateLowerGCFrame::NoteOperandUses(State &S, BBState &BBS, User &UI) {
     for (Use &U : UI.operands()) {
         Value *V = U;
-        if (!isSpecialPtr(V->getType()) && !isSpecialPtrVec(V->getType()))
-            continue;
-        NoteUse(S, BBS, V, Uses);
+        if (isSpecialPtr(V->getType()) || isSpecialPtrVec(V->getType()) ||
```
Comment (Keno) [inline src/llvm-late-gc-lowering.cpp]: I guess, this is the bugfix part of this commit? Add a test? I think `UpExposedUsesUnrooted` was left over from when we still used to optimize that. Also with test, I think this can be a separate PR, so we can get it in quicker.

File: test/llvmpasses/final-lower-gc.ll [hunk 6]
```diff
@@ -28,6 +28,7 @@ top:
 ; CHECK-NEXT: [[GCFRAME_SIZE_PTR2:%.*]] = bitcast %jl_value_t addrspace(10)** [[GCFRAME_SIZE_PTR]] to i64*
 ; CHECK-NEXT: store i64 4, i64* [[GCFRAME_SIZE_PTR2]], !tbaa !0
 ; CHECK-NEXT: [[GCFRAME_SLOT:%.*]] = getelementptr %jl_value_t**, %jl_value_t*** %ptls, i32 0
+; XXX: ordering may be inconsistent, causing the test to fail here on some platforms
```
Comment (Keno) [inline test/llvmpasses/final-lower-gc.ll]: Use `CHECK-DAG`?

**Comment by vtjnash** [comment]
Will merge ~tomorrow, if no comments.

**Comment by StefanKarpinski** [comment]
Keno did leave some comments that appear to be unaddressed. If they're not relevant maybe comment why they're not?
